src/mgos_imu_icm20948.c

```diff
@@ -0,0 +1,482 @@
+    fs = 3;
+    scale = 2000.f;
+  } else {
+    return false;
+  }
+
+  if(!mgos_imu_icm20948_change_bank(dev->i2c, dev->i2caddr, imu_user_data, 2)) {
+    return false;
+  }
+  if (!mgos_i2c_setbits_reg_b(dev->i2c, dev->i2caddr, MGOS_ICM20948_REG2_GYRO_CONFIG_1, 1, 2, fs)) {
+    return false;
+  }
+  dev->opts.scale = scale;
+  dev->scale = dev->opts.scale / 32767.0f;
+
+  return true;
+}
+
+bool mgos_imu_icm20948_gyro_get_odr(struct mgos_imu_gyro *dev, void *imu_user_data, float *odr) {
+  uint8_t div  = 0;
+
+  if(!mgos_imu_icm20948_change_bank(dev->i2c, dev->i2caddr, imu_user_data, 2)) {
+    return false;
+  }
+  if(!mgos_i2c_read_reg_n(dev->i2c, dev->i2caddr, MGOS_ICM20948_REG2_GYRO_SMPLRT_DIV, 1, &div)) {
+    return false;
+  }
+
+  // ODR is computed as follows:
+  // 1.1 kHz/(1+GYRO_SMPLRT_DIV[7:0])
+  *odr = GYRO_BASE_ODR / (div + 1);
+  return *odr >= 0;
+}
+
+bool mgos_imu_icm20948_gyro_set_odr(struct mgos_imu_gyro *dev, void *imu_user_data, float odr) {
+  uint8_t div = 0;
+
+  if(odr <= 0 || odr > GYRO_BASE_ODR) {
+    return false;
+  }
+  div = (GYRO_BASE_ODR / odr) - 1;
+
+  if(!mgos_imu_icm20948_change_bank(dev->i2c, dev->i2caddr, imu_user_data, 2)) {
+    return false;
+  }
+  if(!mgos_i2c_write_reg_b(dev->i2c, dev->i2caddr, MGOS_ICM20948_REG2_GYRO_SMPLRT_DIV, div)) {
+    return false;
+  }
+
+  dev->opts.odr = odr;
+  return true;
+}
+
+bool mgos_imu_icm20948_mag_detect(struct mgos_imu_mag *dev, void *imu_user_data) {
+  int device_id;
+
+  if (!dev->i2c) {
+    return false;
+  }
+
+  device_id = mgos_i2c_read_reg_b(dev->i2c, dev->i2caddr, MGOS_ICM20948_WHO_AM_I_M);
+  if (device_id == MGOS_ICM20948_DEVID_M) {
+    return true;
+  }
+  return false;
+
+  (void)imu_user_data;
+}
+
+bool mgos_imu_icm20948_mag_create(struct mgos_imu_mag *dev, void *imu_user_data) {
+  if (!dev) {
+    return false;
+  }
+
+  // CNTL3: SRST=1;
+  mgos_i2c_write_reg_b(dev->i2c, dev->i2caddr, MGOS_ICM20948_CNTL3_M, 0x01);
+
+  // CNTL2: 01000(MODE4, 100Hz);
+  mgos_i2c_write_reg_b(dev->i2c, dev->i2caddr, MGOS_ICM20948_CNTL2_M, 0x08);
+
+  dev->scale = 22.f / 32768.0;
+  dev->bias[0] = 1.0;
+  dev->bias[1] = 1.0;
+  dev->bias[2] = 1.0;
+  return true;
+
+  (void)imu_user_data;
+}
+
+bool mgos_imu_icm20948_mag_read(struct mgos_imu_mag *dev, void *imu_user_data) {
+  uint8_t data[6];
+
+  if (!dev) {
+    return false;
+  }
+
+  if (!mgos_i2c_read_reg_n(dev->i2c, dev->i2caddr, MGOS_ICM20948_HXL_M, 6, data)) {
+    return false;
+  }
+
+  // It is required to read ST2 register after data reading.
+  mgos_i2c_read_reg_b(dev->i2c, dev->i2caddr, MGOS_ICM20948_ST2_M);
+
+  dev->mx = (data[1] << 8) | (data[0]);
+  dev->my = (data[3] << 8) | (data[2]);
+  dev->mz = (data[5] << 8) | (data[4]);
+  return true;
+
+  (void)imu_user_data;
+}
+
+bool mgos_imu_icm20948_mag_get_odr(struct mgos_imu_mag *dev, void *imu_user_data, float *odr) {
+  int16_t mode = mgos_i2c_read_reg_b(dev->i2c, dev->i2caddr, MGOS_ICM20948_CNTL2_M);
+  if(mode == -1) {
```

Comment:
Please ensure that *odr is a valid pointer before trying to write to it later (causing a crash).
if (!odr) return false;